src/rust/engine/sharded_lmdb/src/lib.rs

```diff
@@ -378,6 +381,109 @@ impl ShardedLmdb {
       .await
   }
 
+  ///
+  /// Stores the given Read instance under its computed digest. This method performs two passes
+  /// over the source to 1) hash it, 2) store it. If !data_is_immutable, the second pass will
+  /// re-hash the data to confirm that it hasn't changed.
+  ///
+  pub async fn store<F, R>(
+    &self,
+    initial_lease: bool,
+    data_is_immutable: bool,
+    data_provider: F,
+  ) -> Result<Digest, String>
+  where
+    R: Read + Debug,
+    F: Fn() -> Result<R, io::Error> + Send + 'static,
+  {
+    let store = self.clone();
+    self
+      .executor
+      .spawn_blocking(move || {
+        let mut attempts = 0;
+        loop {
+          // First pass: compute the Digest.
+          let digest = {
+            let mut read = data_provider().map_err(|e| format!("Failed to read: {}", e))?;
+            let mut hasher = WriterHasher::new(NoopWriter);
+            let _ = std::io::copy(&mut read, &mut hasher)
+              .map_err(|e| format!("Failed to read from {:?}: {}", read, e))?;
+            hasher.finish().0
+          };
+
+          let effective_key = VersionedFingerprint::new(digest.hash, ShardedLmdb::SCHEMA_VERSION);
+          let (env, db, lease_database) = store.get(&digest.hash);
+          let put_res: Result<(), StoreError> = env
+            .begin_rw_txn()
+            .map_err(StoreError::Lmdb)
+            .and_then(|mut txn| {
+              // Second pass: copy into the reserved memory.
+              let dest = txn.reserve(
+                db,
+                &effective_key,
+                digest.size_bytes,
+                WriteFlags::NO_OVERWRITE,
+              )?;
+              let mut read = data_provider().map_err(|e| format!("Failed to read: {}", e))?;
+              if data_is_immutable {
+                // Trust that the data hasn't changed.
+                read
+                  .read_exact(dest)
+                  .map_err(|e| format!("Failed to read from {:?}: {:?}", read, e))?
+              } else {
+                // Confirm that the data hasn't changed.
+                let length_changed = read.read_exact(dest).map(|()| false).or_else(|e| match e
+                  .kind()
+                {
+                  std::io::ErrorKind::UnexpectedEof => {
+                    // If the source was shortened, we should retry rather than erroring.
+                    Ok(true)
+                  }
+                  _ => Err(format!("Failed to read from {:?}: {:?}", read, e)),
+                })?;
+
+                // Should retry if the length changed, or the Digest changed between reads.
+                if length_changed || digest != Digest::of_bytes(dest) {
```

Comment:
It feels like it would be slightly quicker to wrap `dest` in the `read_exact` call in a `HashWriter`, rather than re-reading the whole buffer from the LMDB (which may require reading from disk if we're unlucky with caches)?

Reply:
Yea, potentially... my thinking was that `read_exact` avoided any buffering (read directly from the file and copy into a buffer), but you're right that it could impose more IO if `reserve` is returning us a MMAP'd buffer (...sidenote: I don't think that it is currently because we don't set [WRITE_MAP](https://docs.rs/lmdb/0.8.0/lmdb/struct.EnvironmentFlags.html#associatedconstant.WRITE_MAP): want to look into that next).

Will try both.

Reply:
I tried `WRITE_MAP`. And while it was great in benchmarks (15% faster, low/constant memory usage), it effectively hung when used anywhere other than in macOS's tmpfs... i.e. in `~/.cache/pants/lmdb_store`. So... that's right out!